httpcore/_async/http_proxy.py

```diff
@@ -140,47 +133,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
 
-            # Establish the connection by issuing a CONNECT request...
+            # Issue a CONNECT request...
 
             # CONNECT www.example.org:80 HTTP/1.1
             # [proxy-headers]
             target = b"%b:%d" % (url[1], url[2])
             connect_url = self.proxy_origin + (target,)
-            connect_headers = self.proxy_headers
-            proxy_response = await connection.request(
-                b"CONNECT", connect_url, headers=connect_headers, timeout=timeout
+            headers = self.proxy_headers + ([] if headers is None else headers)
+            proxy_response = await proxy_connection.request(
+                b"CONNECT", connect_url, headers=headers, timeout=timeout
             )
             proxy_status_code = proxy_response[1]
             proxy_reason_phrase = proxy_response[2]
             proxy_stream = proxy_response[4]
 
-            # Ingest any request body.
-            await read_body(proxy_stream)
+            # Read the response data without closing the socket
+            _ = b"".join([chunk async for chunk in proxy_stream])
 
             # If the proxy responds with an error, then drop the connection
             # from the pool, and raise an exception.
             if proxy_status_code < 200 or proxy_status_code > 299:
-                async with self._thread_lock:
-                    self._connections[connection.origin].remove(connection)
-                    if not self._connections[connection.origin]:
-                        del self._connections[connection.origin]
                 msg = "%d %s" % (proxy_status_code, proxy_reason_phrase.decode("ascii"))
                 raise ProxyError(msg)
 
-            # Upgrade to TLS.
-            await connection.start_tls(target, timeout)
+            # Create a new connection to the target
+            connection = AsyncHTTPConnection(
+                origin=origin,
+                http2=False,
+                ssl_context=self._ssl_context,
+                socket=proxy_connection.socket,
+            )
+            async with self._thread_lock:
```

Comment:
Here I'd add a comment about why we're not calling the usual `._add_to_pool()`:

```suggestion
            # Register the connection, but don't count it in pool limits.
            async with self._thread_lock:
```

Reply:
It is the source of a upcoming bug in the HTTPS handling, and very likely a source of other undetected bugs.

I'd prefer not to open another PR just for this though, if you don't mind 🙂

Reply:
You were probably aiming at posting this on https://github.com/encode/httpcore/pull/57#discussion_r407241317? 😅

Reply:
Yep, good catch. It's probably worth adding to the proxy connection as well, if you agree with my comment above.

Reply:
Still not sure about registering the proxy connection, tbh.